lib/core/src/Cardano/Pool/Metadata.hs

```diff
@@ -70,45 +104,129 @@ defaultManagerSettings =
-        -- - A ticker (at most 5 UTF-8 bytes)
-        -- - A homepage (at most 100 UTF-8 bytes)
-        --
-        -- So, the total, including a pretty JSON encoding with newlines ought
-        -- to be less than 512 bytes. For security reasons, we only download the
-        -- first 512 bytes.
-        pure . BL.toStrict <$> brReadSome (responseBody res) 512
-    when (blake2b256 chunk /= hash) $ throwE $ mconcat
+    -> IO (Maybe StakePoolMetadata)
+fetchFromRemote tr builders manager url hash = runExceptTLog $ do
+    chunk <- fromFirst builders
+    when (blake2b256 chunk /= coerce hash) $ throwE $ mconcat
         [ "Metadata hash mismatch. Saw: "
         , B8.unpack $ hex $ blake2b256 chunk
         , ", but expected: "
-        , B8.unpack $ hex hash
+        , B8.unpack $ hex $ coerce @_ @ByteString hash
         ]
     except $ eitherDecodeStrict chunk
   where
-    StakePoolMetadataUrl  url  = url_
-    StakePoolMetadataHash hash = hash_
-
     fromIOException :: Monad m => IOException -> m (Either String a)
     fromIOException = return . Left . ("IO exception: " <>) . show
 
-    fromHttpException :: Monad m => HttpException -> m (Either String a)
-    fromHttpException = return . Left . ("HTTP exception: " <>) . show
+    fromHttpException :: Monad m => HttpException -> m (Either String (Maybe a))
+    fromHttpException = const (return $ Right Nothing)
+
+    runExceptTLog
+        :: ExceptT String IO StakePoolMetadata
+        -> IO (Maybe StakePoolMetadata)
+    runExceptTLog action = runExceptT action >>= \case
+        Left msg ->
+            Nothing <$ traceWith tr (MsgFetchPoolMetadataFailure hash msg)
+
+        Right meta ->
+            Just meta <$ traceWith tr (MsgFetchPoolMetadataSuccess hash meta)
+
+    -- Try each builder in order, but only if the previous builder led to a
+    -- "ConnectionFailure" exception. Other exceptions mean that we could
+    -- likely reach the server.
+    fromFirst (builder:rest) = do
+        uri <- builder url hash
+        req <- requestFromURI uri
+        liftIO $ traceWith tr $ MsgFetchPoolMetadata hash uri
+        mChunk <- ExceptT
+            $ handle fromIOException
+            $ handle fromHttpException
+            $ withResponse req manager $ \res -> do
+            -- NOTE
+            -- Metadata are _supposed to_ be made of:
+            --
+            -- - A name (at most 50 UTF-8 bytes)
+            -- - An optional description (at most 250 UTF-8 bytes)
+            -- - A ticker (at most 5 UTF-8 bytes)
+            -- - A homepage (at most 100 UTF-8 bytes)
+            --
+            -- So, the total, including a pretty JSON encoding with newlines ought
+            -- to be less than 512 bytes. For security reasons, we only download the
+            -- first 512 bytes.
+            if responseStatus res /= status200 then
+                return $ Left "Server did reply, but not positively. This metadata may be blacklisted."
```

Comment:
Does SMASH respond with specific code when metadata is "blacklisted"? If so maybe it could be specified for this particular case.

Currently the solution falls back only if it SMASH url cannot be reached, but for example if provided SMASH url returns `404` it could be perfectly falling back as well...

Reply:
The blacklisting isn't yet implemented in SMASH but, when it does it'll probably return a `404` (according to the specification). There really are two cases:

a) The metadata aren't yet available on SMASH because it hasn't yet downloaded the metadata from a pool (that could be the case for recent pools). Then, it's okay, we'll get a `404` and will retry the url _later_ (exponential backoff), up to several days after the first try and up to 10 times.

b) The metadata is blacklisted, in which case SMASH should also reply with a 404 (very much as if the metadata was absent), and here, we'll retry all-the-same because we can't actually tell the difference with the first case. After 10 retries, we'll stop and consider this metadata "unfetchable". 

When blacklisting is implemented, we might revise this behavior and "acknowledge" blacklisted pools if possible so that we can skip the 10 retries.

Reply:
Then maybe we can be more specific than `/= status200`, and just implement this behaviour on `== status404`? I can imagine that in case of mainetenace, the server might for example return `502` or something else in which case we could just fall back.

Reply:
Also as far as I see SMASH actually responds with 200 a message, in case of data missing:

```
{"code":"DbLookupTxMetadataHash","description":"The hash 534b6a6bd10d23a2eb060a51d61df263bf2fa3f10b6f5a701fd9a77d6c81e67011111111111111111111 is missing from the DB."}
```

Reply:
> Also as far as I see SMASH actually responds with 200 and a message, in case of data missing:

:facepalm: really?

Reply:
https://github.com/input-output-hk/smash/issues/16